src/posit/connect/users.py

```diff
@@ -25,32 +25,73 @@ class User(TypedDict, total=False):
+
+    example:
+    ```python
+
+    from posit.connect import Users
+
+    users = Users(config, session)
+    for user in users:
+        print(user)
+    ```
+
+    Attributes:
+        config (Config): A configuration instance.
+        session (requests.Session): A session instance.
+
+    Methods:
+        find: Returns all users that match a filter.
+        find_one: Returns the first user that matches a filter.
+        get: Returns a user by ID.
+    """
+
+    def __init__(self, config: Config, session: requests.Session) -> None:
+        super().__init__(
+            session, urls.append_path(config.url, "v1/users"), page_size=_MAX_PAGE_SIZE
+        )
         self.config = config
         self.session = session
 
+    # todo - replace with query parameter based filtering.
+    # reason - the current implementation is a fill in for actual implementation
     def find(
-        self, filter: Callable[[User], bool] = lambda _: True, page_size=_MAX_PAGE_SIZE
-    ) -> List[User]:
-        results = Paginator(self.session, self.url, page_size=page_size).get_all()
-        return [User(**user) for user in results if filter(User(**user))]
+        self,
+        filter: Callable[[User], bool] = lambda user: True,
+        *,
+        page_size: int = _MAX_PAGE_SIZE,
+    ) -> list[User]:
+        url = urls.append_path(self.config.url, "v1/users")
+        paginator = Paginator(self.session, url, page_size=page_size)
+        users = (User(**user) for user in paginator)
+        return [user for user in users if filter(user)]
 
+    # todo - replace with query parameter based filtering.
+    # reason - the current implementation is a fill in for actual implementation
     def find_one(
-        self, filter: Callable[[User], bool] = lambda _: True, page_size=_MAX_PAGE_SIZE
-    ) -> User | None:
-        pager = Paginator(self.session, self.url, page_size=page_size)
-        result = pager.get_next_page()
-        while pager.total is None or pager.seen < pager.total:
-            result = pager.get_next_page()
-            for u in result:
-                user = User(**u)
-                if filter(user):
-                    return user
-        return None
+        self,
+        filter: Callable[[User], bool] = lambda user: True,
+        *,
+        page_size: int = _MAX_PAGE_SIZE,
+    ) -> Optional[User]:
+        # You may be tempted to to use self.find(filter, page_size) and return the first element.
+        # This is less efficient as it will fetch all the users and then filter them.
+        # Instead, we use the paginator directly to fetch the first user that matches the filter.
+        # Since the paginator fetches the users in pages, it will stop fetching subsequent pages once a match is found.
+        url = urls.append_path(self.config.url, "v1/users")
+        paginator = Paginator(self.session, url, page_size=page_size)
+        users = (User(**user) for user in paginator)
```

Comment:
Reading this, it's not obvious to me that each of these comprehensions isn't actually iterating over all of the results in the paginator. (There at least used to be versions of Python where that wouldn't be true, unless I'm misremembering.)

So for me, that merits (a) an inline comment acknowledging how it works, and (b) a `responses`-based test that would error if you went beyond a certain page of results. The current test doesn't ensure that, IIRC.

Reply:
I added some comments to try and explain the behavior.